### PR TITLE
elbo with tail adaptive f-divergence

### PR DESCRIPTION
**Pull request by martinjankowiak**
as detailed in this [paper](https://papers.nips.cc/paper/7816-variational-inference-with-tail-adaptive-f-divergence)

invoke as follows: ```TraceTailAdaptive_ELBO(vectorize_particles=True, num_particles=10, tail_adaptive_beta=-1.0)```

### CODE REVIEW

File: pyro/infer/util.py [hunk 1]
```diff
@@ -267,3 +267,11 @@ def compute_expectation(self, costs):
 
         LAST_CACHE_SIZE[0] = count_cached_ops(cache)
         return expected_cost
+
+
+def _check_fully_reparametrized(guide_site):
```
Comment (fritzo) [inline pyro/infer/util.py]: If this is used outside of the file in which it defined, then its name should not begin with underscore.

Comment (martinjankowiak) [reply pyro/infer/util.py]: thanks, done

File: tests/infer/test_inference.py [hunk 2]
```diff
@@ -52,6 +52,9 @@ def test_elbo_reparameterized(self):
     def test_elbo_analytic_kl(self):
         self.do_elbo_test(True, 3000, TraceMeanField_ELBO())
 
+    def test_elbo_tail_adaptive(self):
+        self.do_elbo_test(True, 3000, TraceTailAdaptive_ELBO(num_particles=10, vectorize_particles=True))
```
Comment (fritzo) [inline tests/infer/test_inference.py]: If `TraceTailAdaptive_ELBO` supports nonreparametrized samplers (as your test_valid_models test suggests), then you should add a numerical test that nonreparametrized distributions work. Could you fork this test case?

Comment (martinjankowiak) [reply tests/infer/test_inference.py]: it doesn't support nonreparametrized samplers (as mentioned in docstring). please see test again. unless i'm missing something?

**Comment by martinjankowiak** [comment]
well i think test_subsample_gradient assumes KL divergence, so not applicable

i can add something with a plate though that makes sense

File: tests/infer/test_valid_models.py [hunk 3]
```diff
@@ -1613,16 +1613,22 @@ def guide():
...
         pyro.sample("x", dist.Normal(0., 2.))
 
     def nonrep_guide():
         pyro.sample("x", fakes.NonreparameterizedNormal(0., 2.))
 
-    assert_ok(model, rep_guide, TraceTailAdaptive_ELBO(vectorize_particles=True, num_particles=2))
-    assert_error(model, rep_guide, TraceTailAdaptive_ELBO(vectorize_particles=False, num_particles=2))
-    assert_warning(model, rep_guide, TraceTailAdaptive_ELBO(vectorize_particles=True, num_particles=1))
-    assert_error(model, nonrep_guide, TraceTailAdaptive_ELBO(vectorize_particles=True, num_particles=2))
+    assert_ok(plateless_model, rep_guide, TraceTailAdaptive_ELBO(vectorize_particles=True, num_particles=2))
```
Comment (fritzo) [inline tests/infer/test_valid_models.py]: FYI it's generally useful to split assertions into separate tests so that a failure of the first does not prevent testing of all assertions following. Often we can do this with parametrized tests. Feel free to split this up or not.

File: pyro/infer/trace_tail_adaptive_elbo.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+import torch
+
+from pyro.infer.trace_elbo import Trace_ELBO
+from pyro.infer.util import is_validation_enabled, check_fully_reparametrized
+
+
+class TraceTailAdaptive_ELBO(Trace_ELBO):
+    """
+    Interface for Stochastic Variational Inference with an adaptive
```
Comment (fritzo) [inline pyro/infer/trace_tail_adaptive_elbo.py]: nit: it would be nice to split this into three paragraphs: summary, note on nonreparam, note on loss computation.

File: pyro/infer/elbo.py [hunk 5]
```diff
@@ -45,6 +45,8 @@ class ELBO(object):
         be ignored. Defaults to False.
     :param bool retain_graph: Whether to retain autograd graph during an SVI
         step. Defaults to None (False).
+    :param float tail_adaptive_beta: Exponent beta with -1.0 >= beta > 0.0 for
```
Comment (jpchen) [inline pyro/infer/elbo.py]: `>` -> `<`?

Comment (martinjankowiak) [reply pyro/infer/elbo.py]: whoops

Comment (fritzo) [reply pyro/infer/elbo.py]: Also you'll probably want to enclose these in double ticks
```
Exponent beta with ``-1 <= beta < 0`` for ...
```

Comment (martinjankowiak) [reply pyro/infer/elbo.py]: done

**Comment by martinjankowiak** [comment]
should be good to merge; build error is some unrelated flake
